include/envoy/network/io_handle.h

```diff
@@ -26,8 +25,40 @@ class IoHandle {
   /**
    * Clean up IoHandle resources
    */
-  virtual void close() PURE;
+  virtual Api::SysCallIntResult close() PURE;
+
+  virtual Api::SysCallSizeResult readv(const iovec* iovec, int num_iovec) PURE;
+
+  virtual Api::SysCallSizeResult writev(const iovec* iovec, int num_iovec) PURE;
+
+  virtual bool isClosed() const PURE;
+
+  virtual Api::SysCallIntResult bind(const sockaddr* addr, socklen_t addrlen) PURE;
+
+  virtual Api::SysCallIntResult connect(const struct sockaddr* serv_addr, socklen_t addrlen) PURE;
+
+  virtual Api::SysCallIntResult setSocketOption(int level, int optname, const void* optval,
+                                                socklen_t optlen) PURE;
+
+  virtual Api::SysCallIntResult getSocketOption(int level, int optname, void* optval,
+                                                socklen_t* optlen) const PURE;
+
+  virtual Api::SysCallIntResult getSocketName(struct sockaddr* addr,
+                                              socklen_t* addr_len) const PURE;
+
+  virtual Api::SysCallIntResult getPeerName(struct sockaddr* addr, socklen_t* addr_len) const PURE;
+
+  virtual Api::SysCallIntResult setSocketFlag(int flag) PURE;
+
+  virtual Api::SysCallIntResult getSocketFlag() const PURE;
+
+  virtual Api::SysCallIntResult listen(int backlog) PURE;
+
+  virtual std::unique_ptr<IoHandle> dup() const PURE;
+
+  virtual Api::SysCallIntResult shutdown(int how) PURE;
 };
```

Comment:
@danzh2010 I see is that you have added all socket system calls to the base IoHandle class. What I thought was agreed to in discussions with @alyssawilk and @mattklein123, the whole point of the fd refactoring was to remove the socket interface from the base IoHandle class, and (my interpretation) have generic methods that all derivatives would override. The IoSocketHandle was to handle the details of true sockets, IoVppHandle for a VPPinterface, etc. But here it seems that all IoHandles and the derivatives will look like they are managing a socket interface.

Reply:
I agree this is the intent, but I guess we need to discuss what the generic interface is going to look like. There needs to be some API, and it probably needs to look something like BSD sockets, but should it for example not use `sockaddr` and instead return `Network::Address`, etc.? Probably. I think this is a good time to think about what we want this generic API to look like.

Reply:
I agree that the interface should not be identical to socket system calls, with the same parameters. Furthermore, is it generic enough to have all the virtual methods return SysCallIntResult? Should there be a template class for the return type?

Reply:
Agree that return type should be some template class.
@sbelair2, among the methods I listed in io_handle.h, which do you think VPP will need? Are you concerning that VPP might not want to implement any socket specific API's, i.e. get/setSocketOption? I think two approaches we can have to address you concern if I understood your point correctly:
1. as @mattklein123 suggested, we can still keep all the method that an actual socket will need in IoHandle interface. But change the signature so that they don't look like operating on actual socket. VPP interface can override them with no-op.
2. Take those socket specific method out of IoHandle, and move them to another interface IoSocketHandle which derives from IoHandle. And change the usage of IoHandle in Net:Socket and Network::Address::Instance to  IoSocketHandle. I'm not sure what is the scope of change is going to be. By doing so, any implementation inherits from IoHandle rather than IoSocketHandle won't be able to reuse the code where IoSocketHandle is required, i.e. NetworkSocket.

Right now it's not clear what the general IoHandle interface should have. It might be hard to figure out all the details of the generic methods at this stage. I agree with @alyssawilk that we can figure out the details of each method one at a time.

Reply:
@danzh2010 VPP will need the generic methods of connect, read, write, bind, listen, accept, etc. So your suggestion 1. will basically work if you generalize them somewhat. We ran into this roadblock when I started the PR, putting way too much into the IoHandle interface before it was clear what we needed. But it seems to me that if we are going to support a generic IoHandle interface, the methods required are similar to those of sockets but socket system calls. For example, in the case of VPP, Envoy, through a library, will be talking to a user-level stack and not making system calls. If we want to go with one method at a time I'm fine with that and we can follow @alyssawilk's suggestions.

Reply:
@sbelair2 I have to reconsider the return type of IoHandle interface. If we use template type, all the places where IoHandle is used need to provide a specific type. This defeat the purpose of generalizing all the calls to fd into IoHandle. I guess what we want is to have different return types for different IoHandle implementations. So I propose to use a base class, i.e. IoHandleCallIntResult as return type to replace SysCallIntResult here. 

But we still need to decide what the basic interface such return type should have.  For IoSocketHandle, definitely we want the sys call return value and errno. I could imagine VPP also want at least these two if not more to be returned. 

@mattklein123 Is this PR the right place to have this discussion or we should move to some other place to discuss and document the decision?

Reply:
I think @mattklein123's suggestion is good- shall we continue on slack?

Reply:
I would rather not do slack. Can we either potentially just start a new PR with a proposed interface only and we can comment/iterate on that or also potentially do a gdoc that we can comment on and iterate on? Whichever works best is fine with me.

Reply:
Either works fine for me. @danzh2010 you choose, please.

Reply:
I proposed an interface in #5829. Please give feed back there.